[24.1] Paginate tabular datasets in tables to load invocation reports quicker

ahmedhamidawan opened the pull request:
Tries to improve performance of invocations reports as reported in https://github.com/galaxyproject/galaxy/issues/19439

The `b-table` was crashing the browser tab for tabular datasets with many rows. Using `b-pagination` here to fix that. We were also making a call to fetch datatypes for each dataset in the invocation report. Now we use the store.
Viewing reports is faster now.

https://github.com/user-attachments/assets/d7c18884-6e4c-4cd0-b20e-d28e7c6a18ab

@guerler 's upcoming work on reports in [Overhaul Workflow Reports and Pages #19452](https://github.com/galaxyproject/galaxy/issues/19452) could refactor this so that the `useDatatypesMapper` composable is called from within the `HistoryDatasetDetails` instead of using the store like I did here?

Note: _Targeting 24.1 because the invocation view was introduced then and users clicking on reports can get their whole tab frozen because of this..._

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek reviewed (approved):
That's a nice change, but it doesn't fix the bug. Can you target dev ?

mvdbeek commented on client/src/components/Markdown/Elements/HistoryDatasetDisplay.vue:
```diff
@@ -55,49 +55,46 @@
+                <div v-else-if="error">{{ error }}</div>
+                <div v-else :class="contentClass">
+                    <b-embed
+                        v-if="datatypesMapper?.isSubTypeOfAny(datasetType, ['pdf', 'html'])"
```
The optional chaining here should've been a head-scratcher, it was hiding the real problem of not passing the datatypesMapper in the props.

ahmedhamidawan replied:
To clear up the confusion caused by optional chaining here, should I just add a 
```html
<HistoryDatasetDetails v-else-if="datatypesMapper && ..."
```
to the parent `MarkdownContainer` ?

Actually we need a fix there because we're getting a console error for missing required prop `dataypesMapper`

mvdbeek replied:
Just remove the optional chaining as in the diff, that'll cause plenty of errors to appear in the console. You're already making sure that the mapper is loaded on line 59.

ahmedhamidawan committed:
optimize b-table in `HistoryDatasetDisplay` via pagination

Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>

mvdbeek commented on client/src/components/Markdown/Elements/HistoryDatasetDisplay.vue:
```diff
@@ -54,53 +54,66 @@
+                                        :per-page="perPage"
+                                        :current-page="currentPage"
+                                        :fields="Object.freeze(getFields(metaData))"
+                                        :items="Object.freeze(getItems(itemContent.item_data, metaData))" />
```
```suggestion
                                        :fields="getFields(metaData)"
                                        :items="getItems(itemContent.item_data, metaData)" />
```
I would not optimize things that don't need optimization

mvdbeek commented on client/src/components/Markdown/Elements/HistoryDatasetDisplay.vue:
```diff
@@ -54,53 +54,66 @@
+                                    <b-pagination
+                                        v-model="currentPage"
+                                        align="center"
+                                        :total-rows="Object.freeze(getItems(itemContent.item_data, metaData)).length"
```
```suggestion
                                        :total-rows="getItems(itemContent.item_data, metaData).length"
```

mvdbeek commented on client/src/components/Markdown/Elements/HistoryDatasetDisplay.vue:
```diff
@@ -126,11 +139,17 @@ export default {
         return {
+            currentPage: 1,
             expanded: false,
+            perPage: 200,
```
I would go with 100 ... https://github.com/bootstrap-vue/bootstrap-vue/issues/4155#issuecomment-534815029
```suggestion
            perPage: 100,
```

mvdbeek committed:
Drop Object.freeze optimization

It's not what's causing the slowness, that's entirely because of b-table being slow to render many rows

mvdbeek committed:
Paginate to 100 items per page

We may have many of these displays, so let's be on the safe side here. The embedded frame is less than 100 rows high anyway, so using the pagination control is not a bother anyway.

mvdbeek reviewed (approved):
Thanks @ahmedhamidawan, let's get this in.
